### PR TITLE
Focus prioritized UI in on modal open

### PR DESCRIPTION
**Pull request by stuartromanek**
## Summary

https://linear.app/apostrophecms/issue/PRO-4505/%5Baccessibility%5D-focus-primary-button-in-priority
- modal's `trapFocus` now looks for elements marked `data-apos-focus-priority` and prioritizes it during initial focus setting
- Accessibility improvements for media uploader ui

## Related
- Doc template library module: https://github.com/apostrophecms/doc-template-library/pull/67
- Cypress tests: https://github.com/apostrophecms/testbed/pull/300

## What are the specific steps to test this change?

https://github.com/user-attachments/assets/a1eb3e36-5526-4923-ac18-376d846bb921

## What kind of change does this PR introduce?
*(Check at least one)*

- [ ] Bug fix
- [x] New feature
- [ ] Refactor
- [ ] Documentation
- [ ] Build-related changes
- [ ] Other

### CODE REVIEW

File: modules/@apostrophecms/image/ui/apos/components/AposMediaUploader.vue [hunk 1]
```diff
@@ -237,6 +236,10 @@ export default {
       const isEnterPressed = e.key === 'Enter' || e.code === 'Enter' || e.code === 'NumpadEnter';
       const isSpaceBarPressed = e.keyCode === 32 || e.code === 'Space';
 
+      if (isSpaceBarPressed) {
+        e.preventDefault();
+      }
```
Comment (stuartromanek) [inline modules/@apostrophecms/image/ui/apos/components/AposMediaUploader.vue:241]: Prevents scoll jump

File: modules/@apostrophecms/image/ui/apos/components/AposMediaUploader.vue [hunk 2]
```diff
@@ -1,17 +1,16 @@
 <template>
   <label
     :class="dropzoneClasses"
-    :disabled="disabled"
+    :disabled="disabled ? disabled : null"
```
Comment (stuartromanek) [inline modules/@apostrophecms/image/ui/apos/components/AposMediaUploader.vue:4]: ensures `disabled` is not attributed to element when disabled is `false`

File: modules/@apostrophecms/modal/ui/apos/components/AposModal.vue [hunk 3]
```diff
@@ -268,7 +272,18 @@ function trapFocus() {
...
+
+  focusElement(props.modalData.focusedElement, findPriorityElementOrFirst(elementsToFocus));
+
+  // Components render at various times and can't be counted on to be available on modal's mount
+  // Update the trap focus list until a data-apos-focus-priority element is found or the retry limit is reached
+  if (!foundPriorityElement && findPriorityFocusElementRetryMax.value > currentPriorityFocusElementRetry.value) {
+    await new Promise(resolve => setTimeout(resolve, 200));
+    currentPriorityFocusElementRetry.value++;
+    trapFocus();
+  }
```
Comment (stuartromanek) [inline modules/@apostrophecms/modal/ui/apos/components/AposModal.vue:286]: Simple retry loop to find a priority element. Stops when one has been found or the retry limit has been reached

File: modules/@apostrophecms/image/ui/apos/components/AposMediaUploader.vue (hunk 2 above)
Comment (ETLaurent) [reply modules/@apostrophecms/image/ui/apos/components/AposMediaUploader.vue:4]:
```suggestion
    :disabled="disabled || null"
```

optional

File: modules/@apostrophecms/modal/ui/apos/composables/AposFocus.js [hunk 4]
```diff
@@ -68,7 +69,7 @@ export function useAposFocus() {
 
     const { focusedElement, elementsToFocus } = lastModal;
 
-    focusElement(focusedElement, elementsToFocus[0]);
+    focusElement(focusedElement, findPriorityElementOrFirst(elementsToFocus));
```
Comment (ETLaurent) [inline modules/@apostrophecms/modal/ui/apos/composables/AposFocus.js:72]: 👍

File: modules/@apostrophecms/image/ui/apos/components/AposMediaUploader.vue (hunk 1 above)
Comment (ETLaurent) [reply modules/@apostrophecms/image/ui/apos/components/AposMediaUploader.vue:241]: no side-effects? (when typing `space` in a field for instance)

File: modules/@apostrophecms/modal/ui/apos/components/AposModal.vue [hunk 5]
```diff
@@ -268,7 +272,18 @@ function trapFocus() {
 
   store.updateModalData(props.modalData.id, { elementsToFocus });
 
-  focusElement(props.modalData.focusedElement, props.modalData.elementsToFocus[0]);
+  const firstElementToFocus = findPriorityElementOrFirst(elementsToFocus);
+  const foundPriorityElement = firstElementToFocus.hasAttribute('data-apos-focus-priority');
+
+  focusElement(props.modalData.focusedElement, findPriorityElementOrFirst(elementsToFocus));
```
Comment (ETLaurent) [inline modules/@apostrophecms/modal/ui/apos/components/AposModal.vue:278]:
```suggestion
  focusElement(props.modalData.focusedElement, firstElementToFocus);
```

File: modules/@apostrophecms/modal/ui/apos/components/AposModal.vue [hunk 6]
```diff
@@ -268,7 +272,18 @@ function trapFocus() {
...
+
+  // Components render at various times and can't be counted on to be available on modal's mount
+  // Update the trap focus list until a data-apos-focus-priority element is found or the retry limit is reached
+  if (!foundPriorityElement && findPriorityFocusElementRetryMax.value > currentPriorityFocusElementRetry.value) {
+    await new Promise(resolve => setTimeout(resolve, 200));
+    currentPriorityFocusElementRetry.value++;
+    trapFocus();
```
Comment (ETLaurent) [inline modules/@apostrophecms/modal/ui/apos/components/AposModal.vue:285]: no need to await `trapFocus()` here?

**Review by ETLaurent — COMMENTED** [review]
questions

File: modules/@apostrophecms/modal/ui/apos/components/AposModal.vue [hunk 7]
```diff
@@ -268,7 +272,18 @@ function trapFocus() {
...
-  focusElement(props.modalData.focusedElement, props.modalData.elementsToFocus[0]);
+  const firstElementToFocus = findPriorityElementOrFirst(elementsToFocus);
+  const foundPriorityElement = firstElementToFocus.hasAttribute('data-apos-focus-priority');
+
+  focusElement(props.modalData.focusedElement, findPriorityElementOrFirst(elementsToFocus));
+
+  // Components render at various times and can't be counted on to be available on modal's mount
+  // Update the trap focus list until a data-apos-focus-priority element is found or the retry limit is reached
+  if (!foundPriorityElement && findPriorityFocusElementRetryMax.value > currentPriorityFocusElementRetry.value) {
+    await new Promise(resolve => setTimeout(resolve, 200));
```
Comment (ETLaurent) [inline modules/@apostrophecms/modal/ui/apos/components/AposModal.vue:283]: with 5 retries, it'll wait 1s for elements to be loaded.
is it enough?

File: modules/@apostrophecms/image/ui/apos/components/AposMediaUploader.vue (hunk 1 above)
Comment (stuartromanek) [reply modules/@apostrophecms/image/ui/apos/components/AposMediaUploader.vue:241]: No, is only listening on the uploader UI.
tested
- Uploading a new media item and editing its details in the right-side schema
- Opening the media lib and editing an existing image's details
- Using the search bar

Use of space key was not escaped

File: modules/@apostrophecms/modal/ui/apos/components/AposModal.vue (hunk 7 above)
Comment (stuartromanek) [reply modules/@apostrophecms/modal/ui/apos/components/AposModal.vue:283]: In my testing I haven't seen the retry count go above 2. Do you have a change to suggest? higher retry limit?

File: modules/@apostrophecms/modal/ui/apos/components/AposModal.vue (hunk 6 above)
Comment (stuartromanek) [reply modules/@apostrophecms/modal/ui/apos/components/AposModal.vue:285]: This function doesn't need to wait so I don't think so. But if the best practice would be to await it I can, lmk

**Review by ETLaurent — APPROVED** [review]
lgtm!
haven't tested but let's run Cypress so that we ensure tests are 🆗

**Review by ETLaurent — APPROVED** [review]

File: CHANGELOG.md [hunk 8]
```diff
@@ -4,10 +4,13 @@
...
 
-* Modules can now have a `before: "module-name"` property in their configuration to run (initialization) before another module.
+* Elements inside modals can have a `data-apos-focus-priority` attribute that prioritizes them inside the focusable elements list.
+* Modals will continute trying to find focusable elements until an element marked `data-apos-focus-priority` appears or the max retry threshold is reached.
 
 ### Fixes
-  
+
+* Takes care of an edge case where Media Manager would duplicate search results.
+* Modules can now have a `before: "module-name"` property in their configuration to run (initialization) before another module.
```
Comment (ETLaurent) [inline CHANGELOG.md:13]: should stay in the "### Adds" section

File: modules/@apostrophecms/doc-type/ui/apos/components/AposDocEditor.vue [hunk 9]
```diff
@@ -883,6 +886,12 @@ export default {
       }
 
       return body;
+    },
+    isPriorityButton(name) {
+      const priority = this.restoreOnly ? 'save'
+        : this.saveMenu ? 'splitSave'
+          : this.saveDisabled ? 'cancel' : null;
+      return name === priority || null;
```
Comment (ETLaurent) [inline modules/@apostrophecms/doc-type/ui/apos/components/AposDocEditor.vue:894]: 🤯

**Review by ETLaurent — CHANGES_REQUESTED** [review]
just that.
otherwise if Cypress is 🟢  then it's 🆗

**Review by ETLaurent — APPROVED** [review]

**Pull request merged**